### DISCUSSION TITLE
What are default settings for Java spring/image model for concurrent detections? Is it possible to improve performance on one CPU? (without cluster)

### DISCUSSION THREAD

**Discussion by martinenkoEduard** [exadel-inc/CompreFace#840] [Q&A]
I am not really familiar with java spring - and it seems that REST API and CompreFace Python SDK  works using call to Java spring.
What are default concurrent setting? How many simultaneous detects can it do?
Does it start workers or some analogy for each CPU core? 
What are config files/environment variables that i can tweak ?
Or does it work sequentially? 

When I send several detect requests using CompreFace Python SDK I see that the server's cpu doesn't reach 100% utilization.
Only some core.
So I am curious, is it possible to improve concurrent perfomance. 

And a little bit of repetition but I want to fully understand this:
if I call compreface-python-sdk detect from several separate applications - will they be worked on simultaneously? 
Does compreface work on compreface-python-sd requests at the same time (using workers or threads) or does it use something like a queue and work on the tasks sequentially?

**Comment by martinenkoEduard** [comment]
Does setting several "api_key	string	required	Face Verification Api Key in UUID format" mean that i will increase concurrency? 
Or is it completely irreelevant?

Why do we need different services and different api_keys?

**Reply by martinenkoEduard** [reply]
![Screenshot from 2022-07-24 11-48-08](https://user-images.githubusercontent.com/108512856/180639681-f76a8c63-b5d5-4dda-aa8d-7629c6acf7ae.png)

Does it mean that by default there are several instances?

**Reply by pospielov** [reply]
By default, there are two processes in the CPU build of CompreFace and one process in the GPU build.
According to your screenshot, it looks like there are lots of processes, so it's not the default behavior.

**Comment by VolodymyrBushko** [comment]
What are default concurrent setting?
By default Tomcat/Spring can handle up to 200 parallel requests.


How many simultaneous detects can it do?
The max number of simultaneous detects on the Java side is 200, but you should make sure that the Python side can handle that as well.


Does it start workers or some analogy for each CPU core?
Yes, it does.


What are config files/environment variables that i can tweak?
There are no config files or environment variables that you can tweak the concurrency setting of Tomcat/Spring.


Does it work sequentially?
No, it doesn't.


Does setting several "api_key string required Face Verification Api Key in UUID format" mean that i will increase concurrency?
You can't increase concurrency in that way.
